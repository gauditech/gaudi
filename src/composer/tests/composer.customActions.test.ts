import { compile, compose, parse } from "@src/index";

describe("custom actions", () => {
  it("succeeds for basic composite create", () => {
    const bp = `
    model Org {
      field is_new { type boolean }
      field name { type text }
      reference extras { to OrgExtra, unique }
    }
    model OrgExtra {
      relation org { from Org, through extras }
    }
    entrypoint Orgs {
      target model Org as org
      create endpoint {
        action {
          create OrgExtra as e {}
          create org {
            set is_new true
            set extras e
          }
        }
      }
    }
    `;
    const def = compose(compile(parse(bp)));
    const endpoint = def.entrypoints[0].endpoints[0];
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("fails when default action override is invalid type", () => {
    const bp = `
    model Org {
      field name { type text }
    }
    entrypoint Orgs {
      target model Org
      update endpoint {
        action {
          create {}
        }
      }
    }`;
    const spec = compile(parse(bp));
    expect(() => compose(spec)).toThrowErrorMatchingInlineSnapshot(
      `"Mismatching context action kind: create in endpoint kind: update"`
    );
  });
  it.todo("sets default action if not given");
});
