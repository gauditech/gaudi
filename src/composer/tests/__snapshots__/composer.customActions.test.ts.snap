// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom actions succeeds for basic composite create 1`] = `
[
  {
    "alias": "e",
    "changeset": {},
    "kind": "create-one",
    "model": "OrgExtra",
    "response": [],
  },
  {
    "alias": "org",
    "changeset": {
      "extras": {
        "kind": "reference-value",
        "target": {
          "access": [
            "id",
          ],
          "alias": "e",
        },
        "type": "integer",
      },
      "is_new": {
        "kind": "value",
        "type": "boolean",
        "value": true,
      },
    },
    "kind": "create-one",
    "model": "Org",
    "response": [],
  },
]
`;
